tobias/utils/motifs.py

```diff
@@ -248,70 +278,52 @@ def as_string(self, output_format="pfm"):
-
-				#Convert counts to string
-				counts_type = float if float in [type(element) for element in motif.counts[0]] else int	#Estimate from first base counts
-				if counts_type == float:
-					counts_string = [["{0:.5f}".format(element) for element in row_counts] for row_counts in motif.counts]
-				else:
-					counts_string = [[str(element) for element in row_counts] for row_counts in motif.counts]
-					element_width = max(sum([[len(element) for element in row_counts] for row_counts in counts_string], []))	#Length of longest element to align to
-					counts_string = [[" " * (element_width - len(element)) + element for element in row_counts] for row_counts in counts_string]
-
-				#Write out to format
-				for i, base_counts in enumerate(counts_string):
-					out_string += "{0} [ {1} ]\n".format(bases[i], "  ".join(base_counts)) if output_format == "jaspar" else "  ".join(base_counts) + "\n"
-				out_string += "\n"
+		header = True
+		for motif in self:
+			out_string += motif.as_string(output_format=output_format, header=header)
+			header = False
 
-		elif output_format == "meme":
-			
-			meme_header = "MEME version 4\n\n"
-			meme_header += "ALPHABET= ACGT\n\n"
-			meme_header += "strands: + -\n\n"
-			meme_header += "Background letter frequencies\n"
-			meme_header += " ".join(["{0} {1}".format(bases[i], self.bg[i]) for i in range(4)]) + "\n\n"
-			out_string += meme_header
-
-			for motif in self:
-				out_string += "MOTIF {0} {1}\n".format(motif.id, motif.name)
-				out_string += "letter-probability matrix: "
-				out_string += " ".join("{0}= {1}".format(key, value) for key, value in motif.info.items()) + "\n"
-
-				for i in range(motif.length):
-					row = [float(motif.counts[j][i]) for j in range(4)] 	#row contains original row from content
-					n_sites = round(sum(row), 0)
-					row_freq = ["{0:.6f}".format(num/n_sites) for num in row] 
-					out_string += " " + "  ".join(row_freq) + "\n"
-				
-				out_string += "\n"
-
-		#elif output_format == "transfac":
-		#	for motif in self:
-		#		out_string += motif.to_transfac()
+		return(out_string)
 
-		else:
-			raise ValueError("Format " + output_format + " is not supported")	
+	def get_background(self):
+		"""
+		Combines background of all motifs to a global background.
+  
+		Returns:
+			numpy array of frequencies
+		"""
+		total_n = 0
+		global_bg = None
+  
+		for motif in self:
+			if global_bg is None:
+				global_bg = np.array([0] * len(motif.bg), dtype=float)
 
-		return(out_string)			
+			global_bg += motif.bg * motif.n
+			total_n += motif.n
 
+		return global_bg / total_n
 
 	#---------------- Functions for moods scanning ------------------------#
 
 	def setup_moods_scanner(self):
 		""" Sets self.moods_scanner object """
 
-		tups = [(motif.prefix, motif.strand, motif.pssm, motif.threshold) for motif in self] 		#list of tups
+		# make sure everything necessary exists
+		for motif in self:
+			motif.set_prefix()
+			# creates pssm as well
+			motif.get_threshold(0.05)

```

Comment:
What happened here with the hardcoded pvalue? This dangerously overwrites any previously set pvalue from get_threshold.  Instead of always recalculating the threshold, you could also raise an error if the threshold is still None? 

Same for set_prefix() - maybe you can check first whether the prefix is set, because both have options which would be overwritten with these changes.
```

Reply:
I think I would prefer an error, just because then you don't need to have to deal with the pvalue and prefix strings as parameters to `setup_moods_scanner`. The error could then point out the correct functions for initializing the parameters.

For the saving of the scanner, yes the reason is that there is overhead in calling:
```
scanner = MOODS.scan.Scanner(7)
scanner.set_motifs(self.matrices, self.bg, self.thresholds)
```
So therefore, the scanner is only set once, while `scan_sequence` can be applied to 1000's of sequences using the same scanner.

Reply:
Ok, I will change the for loop to instead check for missing values.